File: E-Commerce-Site/E-Commerce-Site.Users/UsersService.cs
namespace E_Commerce_Site.Users;
using System.Data;
using System.Data.SqlClient;
using Models;

public class UsersService : IUsersService
{
    public async Task AddUser(User user)
    {
        await using SqlConnection connection = new SqlConnection(Helper.CnnVal("EComm"));
        Guid uniqueId = Guid.NewGuid();

        await connection.OpenAsync();
        await using SqlCommand command = new SqlCommand("addUser", connection);
        command.CommandType = CommandType.StoredProcedure;

        command.Parameters.AddWithValue("@userId", uniqueId);
        command.Parameters.AddWithValue("@username", user.Username);
        command.Parameters.AddWithValue("@password", user.Password);

        await command.ExecuteNonQueryAsync();
    }

    public async Task<bool> GetAvailableUser(string username)
    {
        await using SqlConnection connection = new SqlConnection(Helper.CnnVal("EComm"));
        SqlCommand command = new SqlCommand("dbo.[available_username]", connection);
        command.CommandType = CommandType.StoredProcedure;
        command.Parameters.AddWithValue("@username", username);

        await command.Connection.OpenAsync();
        await using SqlDataReader reader = await command.ExecuteReaderAsync();
        var isTaken = await reader.ReadAsync();

        await reader.CloseAsync();
        return !isTaken;
    }
}

File: Python_BackEnd/Users/UsersService.py
from database import get_db_connection
from Users.Models.Users import User
import uuid


# Returns a boolean value representing if username is taken or not
async def get_available_user(username: str) -> bool:
    command = f"EXEC dbo.available_username @username=?"
    params = username
    db = get_db_connection()
    with db.cursor() as cursor:
        cursor.execute(command, params)
        is_taken: bool = cursor.fetchone()
    return not is_taken


# Adds a new user to the database
async def add_user(user: User) -> None:
    user_id = uuid.uuid4()
    command = "EXEC dbo.addUser @userId=?, @username=?, @password=?"
    params = (
        user_id,
        user.username,
        user.password
    )
    db = get_db_connection()
    with db.cursor() as cursor:
        cursor.execute(command, params)


# Login and returns the userID of the user if login successful
async def login(user: User):
    command = "EXEC dbo.login @username=?, @password=?"
    params = (user.username, user.password)
    db = get_db_connection()
    with db.cursor() as cursor:
        cursor.execute(command, params)
        result = cursor.fetchone()
        if result:
            user_id: str = result[0]
            user_uid = uuid.UUID(user_id)
            return user_uid
